<h1 align=center> <font size = 5>CourseRa Capstone Course Assignment - Week 3<font><h1

### Note: This Notebook is representing all three parts of assignment . Please scroll down to review each part or use table of contents.

 ## Table of Contents
 
<div class="alert alert-block alert-info" style="margin-top: 20px">
   
[Part 1. Creating Data Frame From Web Page](#part1)

[Part 2.  Adding Latitude and Longitude for each Neighborhood](#part2)

[Part 3. Explore and Cluster Neighborhoods](#part3)
    
</div>

## <a name="part1">Part 1. Creating Data Frame From Web Page </a>
***

**First step, required libraries will be imported**

In [6]:
# importing required libraries
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

**Lets first transfer table from web page to pandas data frame**

In [7]:
# getting html web page
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
page = requests.get(url)

In [8]:
# parse html content with Beatifulsoup
soup = BeautifulSoup(page.content, 'html.parser')
tb = soup.find('table', class_='wikitable')

In [9]:
# transferring table to pandas data set
neighborhoods = pd.read_html(str(tb))[0]
neighborhoods.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned                        NaN
1         M2A      Not assigned                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [10]:
neighborhoods.shape

(180, 3)

**Now rows with no assigned Borough will be deleted from data set**

In [11]:
# dropping rows with "Not assigned" value in [Borough] column
neighborhoods.drop(neighborhoods[neighborhoods['Borough'] == 'Not assigned' ].index, inplace=True)

neighborhoods.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

**Let's make sure there is no NaN cell in Neighborhood column**

In [12]:
# checking [Neihjborhood] column for "Nan" values
pd.isnull(neighborhoods["Neighborhood"])

2      False
3      False
4      False
5      False
6      False
8      False
9      False
11     False
12     False
13     False
14     False
17     False
18     False
20     False
21     False
22     False
23     False
26     False
27     False
30     False
31     False
32     False
36     False
39     False
40     False
41     False
45     False
46     False
47     False
48     False
       ...  
111    False
112    False
113    False
114    False
116    False
117    False
120    False
121    False
122    False
126    False
129    False
130    False
135    False
138    False
139    False
142    False
143    False
144    False
147    False
148    False
151    False
152    False
153    False
156    False
157    False
160    False
165    False
168    False
169    False
178    False
Name: Neighborhood, Length: 103, dtype: bool

**At the end, Let's check postal column for any duplicate values**

In [13]:
# checking [Postal Code] code for any duplicate values
neighborhoods['Postal Code'].value_counts

<bound method IndexOpsMixin.value_counts of 2      M3A
3      M4A
4      M5A
5      M6A
6      M7A
8      M9A
9      M1B
11     M3B
12     M4B
13     M5B
14     M6B
17     M9B
18     M1C
20     M3C
21     M4C
22     M5C
23     M6C
26     M9C
27     M1E
30     M4E
31     M5E
32     M6E
36     M1G
39     M4G
40     M5G
41     M6G
45     M1H
46     M2H
47     M3H
48     M4H
      ... 
111    M4R
112    M5R
113    M6R
114    M7R
116    M9R
117    M1S
120    M4S
121    M5S
122    M6S
126    M1T
129    M4T
130    M5T
135    M1V
138    M4V
139    M5V
142    M8V
143    M9V
144    M1W
147    M4W
148    M5W
151    M8W
152    M9W
153    M1X
156    M4X
157    M5X
160    M8X
165    M4Y
168    M7Y
169    M8Y
178    M8Z
Name: Postal Code, Length: 103, dtype: object>

In [14]:
neighborhoods.shape

(103, 3)

## <a name="part2">Part 2.  Adding Latitude and Longitude for each Neighborhood</a>
***

### Note: I decied to use provided CSV file to generate longitude and latitude

In [15]:
# getting csv file
!wget -q -O 'Geospatial_Coordinates.csv' https://cocl.us/Geospatial_data

In [22]:
# creating a data frame for postal code
postal_code=pd.read_csv('Geospatial_Coordinates.csv')
postal_code.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [23]:
postal_code.shape

(103, 3)

**Note: Both "Postal_Code" and "neighborhoods" are same size. We can sort both based on Postal Code and merge them**

In [24]:
# sorting "Postal_Code" and "neighborhoods" based [Postal Code] Column
postal_code=postal_code.sort_values(by=['Postal Code'])
neighborhoods=neighborhoods.sort_values(by=['Postal Code'])
neighborhoods.reset_index(drop=True).head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [25]:
# adding [Latitude] and [Longitude] columns to "neighborhoods" data set
neighborhoods['Latitude'] = postal_code['Latitude'].values
neighborhoods['Longitude'] = postal_code['Longitude'].values
neighborhoods.head()

Postal Code      Borough                            Neighborhood  \
9          M1B  Scarborough                          Malvern, Rouge   
18         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek   
27         M1E  Scarborough       Guildwood, Morningside, West Hill   
36         M1G  Scarborough                                  Woburn   
45         M1H  Scarborough                               Cedarbrae   

     Latitude  Longitude  
9   43.806686 -79.194353  
18  43.784535 -79.160497  
27  43.763573 -79.188711  
36  43.770992 -79.216917  
45  43.773136 -79.239476

**Let see how many Borough and Neighborhood exist in final data set**

In [26]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]))

The dataframe has 10 boroughs and 103 neighborhoods.


**Also let see how many neighborhoods are within any Borough**

In [27]:
neighborhoods["Borough"].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
York                 5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

### <a name="part3">Part 3. Explore and Cluster Neighborhoods
***

### Note: There are 103 neighborhoods in data set. For simplicity as recomemded by instructor only neighborhoods with Word "Toronto" will be explored.

**Below Libraries will be required for this section**

In [29]:
import json # library to handle JSON files

!pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    ------------------------------------------------------------
                       

**First, let's create a new data set by extracting neighborhoods for Boroughs containing word "Toronto"**

In [30]:
# selecting neighborhoods with word Toronto in [Borough] Column
toronto_neighborhoods=neighborhoods.loc[(neighborhoods['Borough']== 'Downtown Toronto')|
                                        (neighborhoods['Borough']== 'Central Toronto')|
                                        (neighborhoods['Borough']== 'West Toronto')|
                                        (neighborhoods['Borough']== 'East Toronto')].reset_index(drop=True)
toronto_neighborhoods.shape
toronto_neighborhoods.head()

Postal Code          Borough                    Neighborhood   Latitude  \
0         M4E     East Toronto                     The Beaches  43.676357   
1         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3         M4M     East Toronto                 Studio District  43.659526   
4         M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

**Before moving forward, let's check neighborhoods on Toronto Map**

In [31]:
# getting geographical data for city of Toronto
address = 'Toronto, On'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [45]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_neighborhoods['Latitude'], toronto_neighborhoods['Longitude'], toronto_neighborhoods['Borough'], toronto_neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

**We also need to create credentials to access FourSquare API**

In [46]:
# creating FourSquare API connection
CLIENT_ID = 'H3S1RXV5CSQESSZGHAYJBC52WAW1DCMXL0AIAXZDPDRTMWN5' # your Foursquare ID
CLIENT_SECRET = 'ZWH503IOQO54IXKIGXG0TZBJYW0AQPOXPUJ42PDOYMBMCQCK' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: H3S1RXV5CSQESSZGHAYJBC52WAW1DCMXL0AIAXZDPDRTMWN5
CLIENT_SECRET:ZWH503IOQO54IXKIGXG0TZBJYW0AQPOXPUJ42PDOYMBMCQCK


**We need a function to extract all venues nearby each neighborhood. I used same function in course Lab

In [47]:
#### definning function to extract venues information for all the neighborhoods
def getNearbyVenues(names, latitudes, longitudes, radius, LIMIT):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

**Now let's run above function for Toronto neighborhood data set**

In [48]:
# running above function
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
Toronto_venues = getNearbyVenues(names=toronto_neighborhoods['Neighborhood'],
                                   latitudes=toronto_neighborhoods['Latitude'],
                                   longitudes=toronto_neighborhoods['Longitude'], radius=radius, LIMIT=LIMIT)


The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High Park, The Junction South
Parkdale, Ron

**Let's see how resulting data set looks like**

In [50]:
print (Toronto_venues.shape)
Toronto_venues.head()

(1602, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

**Let's see how many different venues exist for each neighborhood**

In [53]:
toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

print (toronto_onehot.shape)
toronto_onehot.head()


(1602, 235)


Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Toy / Game Store  Trail  Train Station  \
0             0         0  ...                 0      1              0   
1             0         0  ...                 0      0              0   
2             0         0  ...                 0      0              0   
3             0         0  ...                 0      0              0   
4             0         0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wine Shop  Women's Store  Yoga Studio  
0         0          0              0            0  
1         0          0              0            0  
2         0          0              0            0  
3         0          0              0            0  
4         0          0              0            0  

[5 rows x 235 columns]

**We need to know the frequncy of each venue type within neighborhoods**

In [54]:
# calculating frequncy of each venue within neighborhood
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.000000   
2               Business reply mail Processing Centre           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.013699   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12                           Forest Hill North & West           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                                 North Toronto West           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.000000            0.000000      0.000000        0.000000   
2   0.000000            0.000000      0.000000        0.000000   
3   0.058824            0.058824      0.058824        0.117647   
4   0.000000            0.000000      0.000000        0.000000   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.0000

**I'll use same function from lab to determine the most common venue for each neighborhood**

In [55]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

**Now I'll run the above function to create the required data set**

In [56]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2              Business reply mail Processing Centre    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar              Beer Bar                  Café   
1             Nightclub        Breakfast Spot           Coffee Shop   
2            Smoke Shop      Recording Studio  Fast Food Restaurant   
3        Airport Lounge      Airport Terminal           Coffee Shop   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant           Cheese Shop    Seafood Restaurant   
1             Pet Store               Stadium         Burrito Place   
2        Farmers Market         Auto Workshop         Burrito Place   
3       Harbor / Marina                 Plane   Rental Car Location   
4          Burger Joint                   Bar        Ice Cream Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery             Jazz Club                  Hotel  
1            Restaurant          Climbing Gym            Yoga Studio  
2            Restaurant           Pizza Place                Brewery  
3      Sculpture Garden              Boutique          Boat or Ferry  
4   Japanese Restaurant           Salad Place        Bubble Tea Shop

**It's time to start clustering Toronto's neighborhoods**

In [57]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 4, 0, 0, 0, 0], dtype=int32)

In [58]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [59]:
toronto_merged = toronto_neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code          Borough                    Neighborhood   Latitude  \
0         M4E     East Toronto                     The Beaches  43.676357   
1         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3         M4M     East Toronto                 Studio District  43.659526   
4         M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                 Trail                   Pub   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0                  Park  Fast Food Restaurant   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               2                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0     Health Food Store   Distribution Center           Deli / Bodega   
1    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
2            Board Shop     Food & Drink Shop          Sandwich Place   
3             Gastropub                Bakery                 Brewery   
4              Bus Line           Yoga Studio           Deli / Bodega   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0      Department Store                 Dessert Shop                 Diner   
1            Restaurant              Bubble Tea Shop                Bakery   
2          Liquor Store                Burrito Place    Italian Restaurant   
3   American Restaurant                  Yoga Studio    Seafood Restaurant   
4     Electronics Store  Eastern European Restaurant            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store            Yoga Studio  
1                   Pub            Pizza Place  
2            Restaurant         Ice Cream Shop  
3        Sandwich Place            Cheese Shop  
4      Doner Restaurant                Dog Run

**Let's see the results on Map**

In [60]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters